### PR TITLE
Target Android O when building the library

### PR DESCRIPTION
**Pull request by joetimmins**
## Problem
Android O contains significant breaking changes to the way notifications work. It also makes sure that long running services have to be either a) foreground services with a visible notification, or b) killed with extreme prejudice. In order to make sure `download-manager` can handle these differences, we need to compile against Android O.

## Solution
Bump the `compileSdkVersion` and `buildToolsVersion` in the library's build.gradle file to `26`. This causes a load of error toasts to appear when the demo apps are downloading things. In order to fix those, we need to make sure we set a `NotificationChannel` on both the `NotificationManager` we are using and each individual `Notification` we fire through the `NotificationManager`. 

I also added the `NotificationChannelProvider` interface to `DownloadManagerModules` to enable clients of the library to set their own `NotificationChannel`s.

### CODE REVIEW

File: demo-extended/build.gradle [hunk 1]
```diff
@@ -9,13 +9,13 @@ buildscript {
...
-    compileSdkVersion 25
-    buildToolsVersion '25.0.2'
+    compileSdkVersion 26
+    buildToolsVersion '26.0.1'
 
     defaultConfig {
         applicationId "com.novoda.downloadmanager.demo.extended"
         minSdkVersion 14
-        targetSdkVersion 25
+        targetSdkVersion 26
```
Comment (Mecharyry) [inline demo-extended/build.gradle]: @ataulm did you say we should remove the `targetSdkVersion`? I honestly don't remember.

File: library/src/main/java/com/novoda/downloadmanager/DownloadManagerBuilder.java [hunk 2]
```diff
@@ -15,7 +15,6 @@
 /**
  * A client can specify whether the downloads are allowed to proceed by implementing
  * {@link com.novoda.downloadmanager.lib.DownloadClientReadyChecker} on your Application class
- *
```
Comment (Mecharyry) [inline library/src/main/java/com/novoda/downloadmanager/DownloadManagerBuilder.java]: 😆

File: library/src/main/java/com/novoda/downloadmanager/lib/DownloadManagerModules.java [hunk 3]
```diff
@@ -84,6 +88,11 @@ public Builder withNotificationImageRetrieverFactory(@NonNull NotificationImageR
             return this;
         }
 
+        public Builder withNotificationManagerCustomiser(@NonNull NotificationChannelProvider notificationChannelProvider) {
```
Comment (Mecharyry) [inline library/src/main/java/com/novoda/downloadmanager/lib/DownloadManagerModules.java]: The `with` doesn't match the `param`, was there a reason for this?

Comment (joetimmins) [inline library/src/main/java/com/novoda/downloadmanager/lib/DownloadManagerModules.java]: Just an oversight, will fix

File: demo-extended/build.gradle (hunk 1 above)
Comment (joetimmins) [inline demo-extended/build.gradle]: The apps work fine targeting either 25 or 26, I left them as 26 to make the build.gradle files free of warnings.

File: library/src/main/java/com/novoda/downloadmanager/notifications/NotificationDisplayer.java [hunk 4]
```diff
@@ -80,6 +85,19 @@ public Notification buildAndShowNotification(NotificationTag tag, Collection<Dow
         return notification;
     }
 
+    private NotificationCompat.Builder initNotificationBuilder() {
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
+            NotificationChannel notificationChannel = notificationChannelProvider.getNotificationChannel();
+            if (!notificationManager.getNotificationChannels().contains(notificationChannel)) {
```
Comment (Mecharyry) [inline library/src/main/java/com/novoda/downloadmanager/notifications/NotificationDisplayer.java]: Worth adding the `contains` to the `notificationManager`?

Comment (joetimmins) [inline library/src/main/java/com/novoda/downloadmanager/notifications/NotificationDisplayer.java]: We don't own `notificationManager`, it's Google's class

Comment (Mecharyry) [inline library/src/main/java/com/novoda/downloadmanager/notifications/NotificationDisplayer.java]: Ah right, my bad 😄

File: demo-extended/build.gradle (hunk 1 above)
Comment (ataulm) [inline demo-extended/build.gradle]: this is the demo so it's fine (and correct) - it's the core library that we export that should not have target specified

**Pull request merged**